src/stdlib/ListUtils.scillib

```diff
@@ -1,5 +1,66 @@
 library ListUtils
 
+(* Returns result of left fold on f *)
+let list_foldl : forall 'A. forall 'B. ( 'B -> 'A -> 'B) -> 'B -> List 'A -> 'B =
+  tfun 'A => tfun 'B =>
+  fun (f : 'B -> 'A -> 'B) =>
+  let left_f = fun (z: 'B) => fun (x: 'A) =>
+    fun (recurse : 'B -> 'B) => let res = f z x in
+                                recurse res in
+  let foldk = @list_foldk 'A 'B in
+  foldk left_f
+
+(* Returns result of right fold on f *)
+let list_foldr : forall 'A. forall 'B. ('A -> 'B -> 'B) -> 'B -> List 'A -> 'B =
+  tfun 'A => tfun 'B =>
+  fun (f : 'A -> 'B -> 'B) =>
+  let right_f = fun (z: 'B) => fun (x: 'A) =>
+    fun (recurse : 'B -> 'B) => let res = recurse z in f x res in
+  let foldk = @list_foldk 'A 'B in
+  foldk right_f
+
+(* Returns result of folding left on f while Some _ otherwise early termination *)
+let list_foldl_while : forall 'A. forall 'B.
+  ('B -> 'A -> Option 'B) -> 'B -> List 'A -> 'B =
+  tfun 'A => tfun 'B =>
+  fun (f : 'B -> 'A -> Option 'B) =>
+  let whilel_f = fun (z : 'B) => fun(x : 'A) =>
+    fun (recurse : 'B -> 'B) => let res = f z x in
+    match res with
+    | Some acc => recurse acc
+    | None => z
+    end in
+  let foldk = @list_foldk 'A 'B in
+  foldk whilel_f
+
+(* Returns result of folding left on f but terminates early if reaches nth application *)
```

Comment:
1. Can you add the signature for the function?
2. "terminates early if it reaches nth application" somehow seems a bit confusing to me. It may be better to reword as "applies `f` at most `n` times, on consecutive elements of the list" or something like that.

Reply:
The signature I'll add and I can try a new comment and see what everyone prefers.